### PR TITLE
stm32: Configure I2C using new pinctrl via DT

### PR DESCRIPTION
**Pull request by erwango**
This PR converts I2C driver to devicetree pin configuration.

To achieve this, it was required to introduce a new pinctrl helper macro in order
to be able to access pcintrl-x property of a device using its node label (i2c1 for instance)
instead of its instance number as it is done until now in other drivers.

To conform with zephyr devicetree usual helper macro naming scheme:
* Instance based macro was renamed from ST_STM32_DT_PINCTRL to
ST_STM32_DT_INST_PINCTRL (along with its set of internal macros)
* New nodelabel based macro was introduced ST_STM32_DT_PINCTRL

Additionally, to reinforce consistency with other DT_INST macros from zephyr codebase,
I reversed order of arguments in these macros (cf https://github.com/zephyrproject-rtos/zephyr/pull/29057#discussion_r501813511).

Along with this change, existing users of ST_STM32_DT_PINCTRL were updated.

Last, helper macros were provided lengthy documentation.

### CODE REVIEW

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -815,7 +815,7 @@ static void uart_stm32_irq_config_func_##index(const struct device *dev)	\
 STM32_UART_IRQ_HANDLER_DECL(index);					\
 									\
 static const struct soc_gpio_pinctrl uart_pins_##index[] =		\
-					ST_STM32_DT_PINCTRL(0, index);	\
+				ST_STM32_DT_INST_PINCTRL(0, index);	\
```
Comment (gmarull) [inline drivers/serial/uart_stm32.c:818]: `DT_INST*` macros take instance index as a first argument, so I think we should do the same to keep consistency.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c (hunk 1 above)
Comment (erwango) [reply drivers/serial/uart_stm32.c]: Ok, just added a new commit to change this as well.

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml [hunk 2]
```diff
@@ -13,3 +13,13 @@ properties:
...
       required: true
+
+    pinctrl-0:
+      type: phandles
+      required: false
+      description: |
+        GPIO pin configuration for serial signals (SDA, SCL).  We expect
+        that the phandles will reference pinctrl nodes.
+
+        For example the IC1 would be
```
Comment (martinjaeger) [inline dts/bindings/i2c/st,stm32-i2c-v1.yaml:24]:
```suggestion
        For example the I2C1 would be
```

Comment (erwango) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml]: Good catch, fixed

**Review by gmarull — APPROVED** [review]

**Review by martinjaeger — APPROVED** [review]
Looks good. I will update #29037 as soon as this one is merged.

**Review by ABOSTM — APPROVED** [review]

**Pull request merged**